submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,132 @@
+## Support fine-grained cluster connectivity and multiple cable drivers
+
+## Summary
+
+Submariner currently assumes full mesh by default and provision VPN connections between all participating clusters. For
+some use cases, it makes sense to offer finer-grained connectivity options. This could also improve overall
+scalability where  full mesh is not required.
+
+Submariner supports several cable driver implementations (libreswan, wireguard et al) however only one can be
+configured within the same cluster set, that is all clusters that join a broker must use the same cable driver.
+
+There are use cases that involve supporting a mix of cable drivers or cable driver configurations within the same
+cluster set. For example, a cluster set could contain three clusters, two that are on-premise and the other
+in the cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+This could be achieved using different cable driver implementations or the same cable driver implementation with
+differing encryption configurations.
+
+## Proposal
+
+This enhancement proposes a mechanism to allow users to specify their intent as to which clusters to connect and how to
+connect them. The default would still be a full mesh with an homogeneous cable driver.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectivityPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectivityPoliciesSpec `json:"spec"`
+}
+
+type ClusterConnectivityPoliciesSpec struct {
+    Policies []ClusterConnectivityPolicy `json:"policies,omitempty"`
```

Comment:
I believe it's more standard in k8s to have separate policies, instead of a big one with all policies.

As you explained this is to be specific about priorities order, and we talked about not using prioirty numbers (magic numbers), and may be another object where we can provide the order of policies when that needs to be resolved. I guess the spec wasn't still updated since last meeting.

Reply:
A good solution for these concerns can be:
1. separate policies, without a big list of policies, while enforcing a unique naming scheme for policies
2. add a PolicyOrder element that will list the policies by name from the most preferred to the least preferred, here we can also specify the "default" policy (for example - this can include the default cable driver to use).  Although YAML does not support ordered lists directly we can easily use a dictionary for this task - which will also make this added order policy more readable.

The clear benefits of this approach:
1. we can easily figure out the exact order of priorities and scan to understand which cluster set uses which policy
2. we achieve flexibility within the YAML
3. we can generate complex policies with modular and clean policy definitions  - for example: the policy of "clients can access the server but not each other" can be written *below* (in the PolicyOrder) the policy of "Admin clients in each cluster set can speak to each other with a secured connection" - This generates the wanted policy "clients should only talk with the server unless they are Admin clients who can talk to each other but only in a secure manner" in a modular, clean and easy way.

One downside to this approach is that we will not have a single place that specify the entire policy rather a scattered partially written rules. An approach to handle this is to issue a log/ print/ unique config file with the final policies together - every connection and its specs.

Reply:
There's pros and cons with both approaches. I'm fine with adding a `ClusterConnectivityPolicyOrder` resource. One downside with it is that a referenced policy in the list can become stale if it's deleted and not removed from the list by the user but perhaps we can do that automatically.

Reply:
I like the cluster connectivity policy order approach. Provides more flexibility, although it has the cons of more computation and being less atomic.

Reply:
But honestly I’m also good with the current single object approach. 

The advantage of splitting is if we believe that this policy could also be targeted by multiple workers creating or modifying policies in real time (less
contention over a single object) but I cannot envision such case.

so the current form is possibly good.